doc: nrf: libraries: others: Add nRF compression documentation (#17150)
nrfconnect/sdk-nrf: doc/nrf/libraries/others/nrf_compression.rst

.. _nrf_compression:

nRF Compression
###############

The nRF compression library in |NCS| provides a streamlined API for providing functions to compress and decompress data on devices.
Different compression types can be enabled by setting Kconfig options.
This library can be enabled with :kconfig:option:`CONFIG_NRF_COMPRESS`, compression is not currently supported, decompression can be enabled with :kconfig:option:`CONFIG_NRF_COMPRESS_DECOMPRESSION`.
The following compression types are avilable and can be enabled using the corresponding Kconfigs:

+------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+

> | Name             | Kconfig                                                                                                 | Notes                                                             |

Author: it's not properties or specification, these are notes about the implementation in general, not actual limits from a specification

> | Name             | Kconfig                                                                                                 | Notes                                                             |

Reviewer: Got it, thanks! Would "implementation details" be the right way to categorize it then?

> | Name             | Kconfig                                                                                                 | Notes                                                             |

Author: That works, will update

| LZMA version 1   | :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA` and :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA1` | Exclusive/cannot be enabled with LZMA version 2,                  |
|                  |                                                                                                         | Fixed probability size of 14272B, fixed dictionary size of 128KiB |
+------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+
| LZMA version 2   | :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA` and :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2` | Exclusive/cannot be enabled with LZMA version 1,                  |
|                  |                                                                                                         | Fixed probability size of 14272B, fixed dictionary size of 128KiB |
+------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+

> | ARM thumb filter | ``NRF_COMPRESS_ARM_THUMB``                                                                              |                                                                   |

Author: It is in an open PR which has not been merged yet

Defining compression type
-------------------------

Once the code has been created, the library needs to be defined properly in an iterable section by using the :c:macro:`NRF_COMPRESS_IMPLEMENTATION_DEFINE` macro, this is defined in the header file :zephyr_file:`include/nrf_compress/implementation.h`, if a library only supports compression then the compression function must be defined and two decompression functions set to ``NULL``, if a library only supports decompression then the two decompression functions must be defined and compression function set to ``NULL``, if both are supported then all three compression/decompression functions must be defined, all other functions are always mandatory and must always be defined.
A unique ID for the compression library also needs to be defined, this should be globally available so that any applications or modules wanting to use this compression library can use it, as that is how compression libraries are indexed.

Using the compression subsystem
*******************************

> The following steps are required to decompress data using the nRF compression library (a flowchart diagram follows this description):
> 
> * :c:func:`nrf_compress_implementation_find` should be used with the ID of the compression type to find, if this returns NULL then the compression type is not supported.
> * The ``init``, ``deinit`` and ``reset`` core function pointers should be checked, if any are NULL then there is an implementation problem with the compression type.
> * For decompression support the ``decompress_bytes_needed`` and ``decompress`` function pointers should be checked, if either are NULL then there is an implementation problem with the compression type.
> * The ``init`` function should be called to set the compression library up, if a non-zero error is returned then the compression library cannot be initialized, this could be due to lack of sufficient memory or other issues.
> * The ``decompress_bytes_needed`` function should be called to see how many bytes of data should be specified for decompression, if a non-zero error is returned then there is an issue with the compression library.
> * The ``decompress`` function should be called with the requested amount of data (or the maximum amount of data if there is less than that), if there is no more data to process then the ``last_part`` value should be set to true, otherwise false. if a non-zero error is returned then it is likely that the data provided to decompress is not valid.
> * If the ``offset`` updated value is not equal to the size of the input data provided then the next call to ``decompress`` should include the unused bytes from the buffer at the start of it.
> * If the ``output_size`` value is greater than 0 then the amount of data specified is available in the ``output`` buffer which should be copied or moved to its intended destination.
> * The process of calling ``decompress_bytes_needed`` followed by ``decompress`` should continue until all the data has been processed.
> * The ``deinit`` function should be called to set clean the compression library up.

Reviewer: ```suggestion
To decompress data using the nRF Compression library, complete the following steps:

1. Use the :c:func:`nrf_compress_implementation_find` function with the ID of the desired compression type. 
   If it returns ``NULL``, the compression type is not supported.
#. Check the ``init``, ``deinit`` and ``reset`` core function pointers.
   If any return ``NULL``, there is an implementation issue with the compression type.
#. For decompression support, check the :c:type:`nrf_compress_decompress_bytes_needed_t` and :c:func:`nrf_compress_decompress_func_t` function pointers. 
   If either is ``NULL``, there is an implementation issue with the compression type.
#. Call the :c:func:`nrf_compress_init_func_t` function to set up the compression library.
   If a non-zero error is returned, the library cannot be initialized, possibly due to insufficient memory or other issues.
#. Call the :c:type:`nrf_compress_decompress_bytes_needed_t` function to determine how many bytes of data should be specified for decompression.
   If a non-zero error is returned, there is an issue with the compression library.
#. Call the :c:func:`nrf_compress_decompress_func_t` function with the requested amount of data (or the maximum available, if less is available). 
   In case there is no more data to process, set the ``last_part`` value to true, otherwise set it to false. 
   If a non-zero error is returned, the provided data might be invalid.
#. Ensure that if the ``offset`` updated value is not equal to the size of the input data provided, the next call to  :c:func:`nrf_compress_decompress_func_t` function includes the unused bytes from the buffer at its start.
#. Verify if the ``output_size`` value is greater than ``0``.
   In such a case, the specified amount of data is available in the ``output`` buffer and you should copy it or move it to its intended destination.
#. Repeat the process of calling the :c:type:`nrf_compress_decompress_bytes_needed_t` function followed by  :c:func:`nrf_compress_decompress_func_t` until all the data has been processed.
#. Call the :c:func:`nrf_compress_deinit_func_t` function to clean up the compression library.
```

.. figure:: nrf_compression_image.png
    :alt: nRF compress library decompression flowchart

MCUboot image compression
#########################

> MCUboot in |NCS| optionally supports compressed image updates, the features and limitations of this system include:
> 
> * Allows slot1 to be smaller than slot0 (ideally slot1 being 70% of the size of slot0).
> * Supports a single image only (does not support network core updates for nRF5340 or NSIB updates).
> * Does not support reverts so MCUboot must be configured for upgrade-only mode.
> * Compressed image will be validated during update process before main image is erased to ensure it decompresses without error, without leading to a bricked module due to no loadable firmware.
> * Image encryption is not supported.
> * LZMA2 compression with ARM thumb filter for compressed images.
> * Static partition manager file is required.
> * Sysbuild must be used.

Reviewer: ```suggestion
MCUboot in the |NCS| optionally supports compressed image updates. 
The system includes the following features and limitation:

* Allows slot 1 to be approximately 70% the size of slot 0.
* Supports a single image only.
  It does not support network core updates for the nRF5340 SoC or NSIB updates.
* Does not support reverting to previous versions. 
  MCUboot must be configured for upgrade-only mode.
* Validates the compressed image during the update process before the main image is erased, ensuring the update does not lead to a bricked module due to un-loadable firmware. 
* Does not support image encryption.
* Uses LZMA2 compression with ARM thumb filter for compressed images.
* Requires a static Partition Manager file.
* Must use sysbuild.
```

.. _nrf_compression:

nRF Compression
###############

The nRF Compression library in the |NCS| provides a streamlined API for data compression and decompression on devices.
You can enable different compression types by configuring the appropriate Kconfig options.
To enable this library, set the :kconfig:option:`CONFIG_NRF_COMPRESS` Kconfig option.
For decompression, set the :kconfig:option:`CONFIG_NRF_COMPRESS_DECOMPRESSION` Kconfig option.
Compression functionality is not currently supported.

You can use the available compression types by enabling their respective Kconfig options:

> +------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+
> | Name             | Kconfig                                                                                                 | Notes                                                             |
> +==================+=========================================================================================================+===================================================================+
> | LZMA version 1   | :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA` and :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA1` | Exclusive/cannot be enabled with LZMA version 2.                  |
> |                  |                                                                                                         | Fixed probability size of 14272 bytes. Fixed dictionary size of 128 KiB |
> +------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+
> | LZMA version 2   | :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA` and :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2` | Exclusive/cannot be enabled with LZMA version 1.                  |
> |                  |                                                                                                         | Fixed probability size of 14272 bytes. Fixed dictionary size of 128 KiB |
> +------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+
> | ARM thumb filter | ``NRF_COMPRESS_ARM_THUMB``                                                                              |                                                                   |
> +------------------+---------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------+

Reviewer: ```suggestion
+------------------+---------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------+
| Name             | Kconfig                                                                                                 | Implementation details                                                             |
+==================+=========================================================================================================+====================================================================================+
| LZMA version 1   | :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA` and :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA1` | Exclusive/cannot be enabled with LZMA version 2.                                   |
|                  |                                                                                                         | Fixed probability size of 14272 bytes. Fixed dictionary size of 128 KiB            |
+------------------+---------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------+
| LZMA version 2   | :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA` and :kconfig:option:`CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2` | Exclusive/cannot be enabled with LZMA version 1.                                   |
|                  |                                                                                                         | Fixed probability size of 14272 bytes. Fixed dictionary size of 128 KiB            |
+------------------+---------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------+
| ARM thumb filter | ``NRF_COMPRESS_ARM_THUMB``                                                                              | --                                                                                 |
+------------------+---------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------+
```
This fixes the table build error with new heading

Compression and decompression can use a significant amount of memory.
To manage this, use the following Kconfig options to choose between static and dynamic (malloc) allocations from the system heap:

* :kconfig:option:`CONFIG_NRF_COMPRESS_MEMORY_TYPE_STATIC` - This is the default option that uses static buffers, ensuring their availability but preventing other uses of the memory.
* :kconfig:option:`CONFIG_NRF_COMPRESS_MEMORY_TYPE_MALLOC` - The option uses dynamic memory allocation, requiring the heap to have sufficient contiguous free memory for buffer allocation upon initializing the compression type.
  This allows other parts of the application to utilize the memory when the compression system is not in use.

> Other Kconfig options for this subsystem include:

Reviewer: ```suggestion
Other Kconfig options for this subsystem include:

```
to fix the build error

#. Call the :c:func:`nrf_compress_decompress_func_t` function with the requested amount of data (or the maximum available, if less is available).
   In case there is no more data to process, set the ``last_part`` value to true, otherwise set it to false.
   If a non-zero error is returned, the provided data might be invalid.
#. Ensure that if the ``offset`` updated value is not equal to the size of the input data provided, the next call to  :c:func:`nrf_compress_decompress_func_t` function includes the unused bytes from the buffer at its start.
#. Verify if the ``output_size`` value is greater than ``0``.
   In such a case, the specified amount of data is available in the ``output`` buffer and you should copy it or move it to its intended destination.
#. Repeat the process of calling the :c:type:`nrf_compress_decompress_bytes_needed_t` function followed by  :c:func:`nrf_compress_decompress_func_t` until all the data has been processed.
#. Call the :c:func:`nrf_compress_deinit_func_t` function to clean up the compression library.

.. figure:: nrf_compression_image.png
    :alt: nRF compress library decompression flowchart

MCUboot image compression

> #########################

Reviewer 2: This header level is only for page titles. I suggest the following:
```suggestion
*************************
```
And then changing the following subsection underlines accordingly to one level lower.

Author: Should clarify that this will be a different page, I've got it in one file at present just to see it in the documentation output but it will be moved to a different file once finished

MCUboot in the |NCS| optionally supports compressed image updates.
The system includes the following features and limitation:

* Allows slot 1 to be approximately 70% the size of slot 0.
* Supports a single image only.
  It does not support network core updates for the nRF5340 SoC or NSIB updates.
* Does not support reverting to previous versions.
  MCUboot must be configured for upgrade-only mode.
* Validates the compressed image during the update process before the main image is erased, ensuring the update does not lead to a bricked module due to un-loadable firmware.
* Does not support image encryption.
* Uses LZMA2 compression with ARM thumb filter for compressed images.
* Requires a static Partition Manager file.
* Must use sysbuild.

Sample

A sample demonstrating compressed MCUboot image support is offered in TODO.

Setup

> Partition manager
> *****************

Reviewer 2: ```suggestion
Static Partition Manager file for MCUboot
=========================================
```

The static Partition Manager file should include the following partitions:

>   * boot_partition: Requires a minimum size of 48 KiB for a minimal build without logging.
>   * slot1_partition: Should be approximately 70% of the size of slot 0 partition for optimal configuration, assuming that image savings will be 30%.
>     The total compression depends on the data within the image.

Reviewer 2: ```suggestion
* ``boot_partition``: Requires a minimum size of 48 KiB for a minimal build without logging.
* ``slot1_partition``: Should be approximately 70% of the size of slot 0 partition for optimal configuration, assuming that image savings will be 30%.
    The total compression depends on the data within the image.
```
